Sources/Foundation/FileHandle+Async.swift

```diff
@@ -0,0 +1,189 @@
+    // The reasons it's delicately safe here are:
+    // • We never use the pointer to access a property (would violate exclusivity)
+    // • We never access the interior of a value type (doesn't have a stable address)
+    //     - This is especially delicate in the case of Data, where we have to force it out of its inline representation
+    //       which can't be reliably done using public API
+    // • We keep the reference we're accessing the interior of alive manually
+    var baseAddress: UnsafeMutableRawPointer {
+        (storage as! Storage).withUnsafeMutablePointerToElements { UnsafeMutableRawPointer($0) }
+    }
+    
+    var capacity: Int {
+        (storage as! Storage).capacity
+    }
+    
+    var storage: AnyObject? = nil
+    @usableFromInline internal var nextPointer: UnsafeMutableRawPointer
+    @usableFromInline internal var endPointer: UnsafeMutableRawPointer
+    
+    @usableFromInline init(capacity: Int) {
+        let s = Storage.create(minimumCapacity: capacity) { _ in
+            return Header(readFunction: nil, finished: false)
+        }
+        storage = s
+        nextPointer = s.withUnsafeMutablePointerToElements { UnsafeMutableRawPointer($0) }
+        endPointer = nextPointer
+    }
+    
+    @inline(never) @usableFromInline
+    internal mutating func reloadBufferAndNext() async throws -> UInt8? {
+        let storage = self.storage as! Storage
+        if storage.finished {
+            return nil
+        }
+        try Task.checkCancellation()
+        nextPointer = storage.withUnsafeMutablePointerToElements { UnsafeMutableRawPointer($0) }
+        do {
+            let readSize = try await readFunction(&self)
+            if readSize == 0 {
+                storage.finished = true
+            }
+        } catch {
+            storage.finished = true
+            throw error
+        }
+        return try await next()
+    }
+    
+    @inlinable @inline(__always)
+    internal mutating func next() async throws -> UInt8? {
+        if _fastPath(nextPointer != endPointer) {
+            let byte = nextPointer.load(fromByteOffset: 0, as: UInt8.self)
+            nextPointer = nextPointer + 1
+            return byte
+        }
+        return try await reloadBufferAndNext()
+    }
+}
+
+extension FileHandle {
+    
+    @available(macOS 12.0, iOS 15.0, tvOS 15.0, watchOS 8.0, *)
+    public struct AsyncBytes: AsyncSequence {
+        public typealias Element = UInt8
+        public typealias AsyncIterator = FileHandle.AsyncBytes.Iterator
+        var handle: FileHandle
+        
+        internal init(file: FileHandle) {
+            handle = file
+        }
+        
+        public func makeAsyncIterator() -> Iterator {
+            return Iterator(file: handle)
+        }
+        
+        @frozen
+        public struct Iterator: AsyncIteratorProtocol {
+            
+            @inline(__always) static var bufferSize: Int {
+                16384
```

Comment:
It occurs to me this should probably be 16384 minus the size of the object header

Reply:
Or maybe not so the loop can be aligned. Ugh, I'll need to dig into this at some point.